docs/root/version_history/current.rst

```diff
@@ -30,6 +30,9 @@ Minor Behavior Changes
   ``envoy.reloadable_features.send_strict_1xx_and_204_response_headers``
   (do not send 1xx or 204 responses with these headers). Both are true by default.
 * http: serve HEAD requests from cache.
+* http: set the default lazy headermap threshold to 3, which defines the minimal number of headers in a
```

Comment:
Thanks for improving the defaults - I like removing O(n) problems from the default deployment :-)

I think it's not clear from this what the user-visible change is likely to be.  I assume there's going to be some user-visible change or we'd just have done a runtime guard of "use_efficient_header_map" and planned to deprecate the old code. 

Is it the case that if there's more than 3 non-inline headers, that Envoy will elide multiple header keys into a single entry?

Reply:
+1 on making this more clear, and potentially even adding a section on this to the HTTP arch overview? 

Stepping back, I feel like if we are going to keep this configurability it should probably be a config option or a CLI option, but that would be a larger change. At minimum I would recommend more documentation.

Reply:
Nice, can you ref link to the new section you added in the arch overview docs? (You will need to add an anchor above that new section.)

Reply:
sorry I guess I'm still not clear on the user-visible change
Does this just add an extra data structure for fast look-up, but ordering and inlining will remain the same?  If so why aren't we just turning this on by default, seeing if anyone complains, and removing the old option?  Basically any request and response headers are going to go above 3, so this seems like a lot of complexity for a feature that'd always be on unless someone turns it off to make their own operations more expensive? :-P

Reply:
I tend to agree that if the actual number we are picking is 3 (as opposed to 10 for example), having it always on and hard coded internally is probably OK. It wouldn't turn on for many trailers cases but that is a minor point.

Reply:
ah, trailers is a good counterpoint - often only one or two of those. might be worth commenting somewhere the value was selected to result in doing fast mapping for the one and not the other.   I wonder if this catches most 100-continue or if they pass through as well. either way feel free to merge once CI passes.